_utils/TileCodingAgent.py

```diff
@@ -0,0 +1,202 @@
+import numpy as np
+from typing import Any, Dict, List
+from PyExpUtils.utils.random import sample
+from ReplayTables.interface import Timestep
+from ReplayTables.ingress.LagBuffer import LagBuffer
+from RlGlue import BaseAgent
+from numba import njit
+from PyExpUtils.utils.jit import try2jit
+from PyFixedReps.TileCoder import TileCoder, TileCoderConfig
+
+
+
+@njit(cache=True)
+def _update(w, x, a, xp, pi, r, gamma, alpha):
+    qsa = w[a][x].sum()
+
+    qsp = w.T[xp].sum(axis=0)
+
+    delta = r + gamma * qsp.dot(pi) - qsa
+
+    w[a][x] = w[a][x] + alpha / len(x) * delta
+
+
+@njit(cache=True)
+def value(w, x):
+    return w.T[x].sum(axis=0)
+
+
+class SparseTileCoder(TileCoder):
+    def __init__(self, params: TileCoderConfig, rng=None):
+        params.scale_output = False
+        super().__init__(params, rng=rng)
+
+    def encode(self, s):
+        return super().get_indices(s)
+
+
+@try2jit
+def argsmax(arr: np.ndarray):
+    ties: List[int] = [
+        0 for _ in range(0)
+    ]  # <-- trick njit into knowing the type of this empty list
+    top: float = arr[0]
+
+    for i in range(len(arr)):
+        if arr[i] > top:
+            ties = [i]
+            top = arr[i]
+
+        elif arr[i] == top:
+            ties.append(i)
+
+    if len(ties) == 0:
+        ties = list(range(len(arr)))
+
+    return ties
+
+
+@njit(cache=True)
+def egreedy_probabilities(qs: np.ndarray, actions: int, epsilon: float):
+    # compute the greedy policy
+    max_acts = argsmax(qs)
+    pi: np.ndarray = np.zeros(actions)
+    for a in max_acts:
+        pi[a] = 1.0 / len(max_acts)
+
+    # compute a uniform random policy
+    uniform: np.ndarray = np.ones(actions) / actions
+
+    # epsilon greedy is a mixture of greedy + uniform random
+    return (1.0 - epsilon) * pi + epsilon * uniform
+
+
+class TileCodingAgent(BaseAgent):
+    def __init__(self):
+        ####################################################
+        # TODO: bad practice, should later figure out what to do to avoid hard coding these
```

Comment:
Since it's just for testing purposes, I'm actually quite comfortable with just having a bunch of hardcoded hypers here! Basically, the more "irrelevant" degrees of freedom that we can eliminate, the better our tests will be.